File: C++/Unit6/exp7_4.cpp
#include <iostream>
using namespace std;

class Person
{
    char name[20];
    char sex[5];
    char id_card[20];
    char birth[20];

    public:
        Person(char* name, char* sex, char* id_card, char* birth)
        {
            strcpy(this->name, name);
            strcpy(this->sex, sex);
            strcpy(this->id_card, id_card);
            strcpy(this->birth, birth);
        }
        
        void Show()
        {
            // cout<<"信息又:<<endl;
            cout<<"姓名:"<<name<<endl;
            cout<<"性别:"<<sex<<endl;
            cout<<"生份证号:"<<id_card<<endl;
            cout<<"生日:"<<birth<<endl;
        }
};


class Student: virtual public Person
{
    char no[20];
    int  grade;

    public:
        Student(char* no, int grade, char* name, char* sex, char* id_card, char* birth):Person(name, sex, id_card, birth)
        {
            strcpy(this->no, no);
            this->grade = grade;
        }

        void Show()
        {
            // cout<<"学生信息有:"<<endl;
            // Person::Show();
            cout<<"学生学号:"<<no<<endl;
            cout<<"学生成绩:"<<grade<<endl;
        }
};


class Teacher:virtual public Person
{
    char position[10];

    public:
        Teacher(char* pos, char* name, char* sex, char* id_card, char* birth):Person(name, sex, id_card, birth)
        {
            strcpy(this->position, pos);
        }

        void Show()
        {
            // cout<<"教师信息:"<<endl;
            // Person::Show();
            cout<<"职位:"<<position<<endl;
        }
};


class Stu_Tech: public Student, public Teacher
{
    public:
        Stu_Tech(char* pos, char* no, int grade, char* name, char* sex, char* id_card, char* birth)
        :Person(name, sex, id_card, birth), Teacher(pos, name, sex, id_card, birth), Student(no, grade, name, sex, id_card, birth)
        {
            cout<<"stu_tea constructing..."<<endl;
        }

        void Show()
        {
            cout<<"stu_teacher info:"<<endl;
            Person::Show();
            Student::Show();
            Teacher::Show();
        }
};

int main()
{
    class Stu_Tech St("讲师", "2020", 100, "hello", "男", "3838", "0501");
    St.Show();
}

File: C++/Unit6/go/exp7_4.go
package main

import "fmt"

// 人 基类
type Person struct {
	name    string
	sex     string
	id_card string
	birth   string
}

func (temp Person) Show() {
	fmt.Println("姓名:", temp.name)
	fmt.Println("性别:", temp.sex)
	fmt.Println("生份证号:", temp.id_card)
	fmt.Println("生日:", temp.birth)
}

// 学生 派生类
type Student struct {
	Person
	no    string
	grade int
}

func (temp Student) Show() {
	// temp.Show()
	fmt.Println("学号:", temp.no)
	fmt.Println("成绩:", temp.grade)
}

// 老师 派生类
type Teacher struct {
	Person
	position string
}

func (temp Teacher) Show() {
	println("职位:", temp.position)
}

func main() {
	mike := Person{
		name:    "mike",
		sex:     "男",
		id_card: "2020213791",
		birth:   "2020-5-1",
	}
	mike.Show()

	println()

	cyw := Student{Person{"cyw", "girl", "363401", "2002"}, "2020213", 100}

	cyw.Person.Show()
	cyw.Show()

	println()

	zk := Teacher{Person{"zk", "boy", "360802", "2002"}, "教授"}
	zk.Person.Show()
	zk.Show()
}
